src/buildstream/source.py

```diff
@@ -696,14 +699,22 @@ def get_mirror_directory(self) -> str:
 
         return self.__mirror_directory
 
-    def translate_url(self, url: str, *, alias_override: Optional[str] = None, primary: bool = True) -> str:
+    def translate_url(
+        self,
+        url: str,
+        *,
+        alias_override: Optional[str] = None,
+        primary: bool = True,
+        extra_data: Optional[Dict[str, Any]] = None,
+    ) -> str:
         """Translates the given url which may be specified with an alias
         into a fully qualified url.
 
         Args:
            url: A URL, which may be using an alias
            alias_override: Optionally, an URI to override the alias with.
            primary: Whether this is the primary URL for the source.
+           extra_data: Additional data provided by :class:`SourceMirror <buildstream.sourcemirror.SourceMirror>` (*Since: 2.2*)
```

Comment:
Shall we provide guidance how a source plugin should deal with this. I would say source plugins should fail if the mirror provides a key it doesn't understand. Or is this too restrictive? `DownloadableFileSource` in #1895 currently doesn't fail in that case, though.